// @antv/util/lib/array.js
var isArrayLike = require('./type/isArrayLike');

var arrayProto = Array.prototype;
var slice = arrayProto.slice;
var indexOf = arrayProto.indexOf;
var splice = arrayProto.splice;

function contains(arr, value) {
  if (!isArrayLike(arr)) {
    return false;
  }
  return indexOf.call(arr, value) > -1;
}

function uniq(arr) {
  var resultArr = [];
  arr.forEach(function (item) {
    if (!contains(resultArr, item)) {
      resultArr.push(item);
    }
  });
  return resultArr;
}

function pull(arr) {
  /**
   * const arr = [ 'a', 'b', 'c', 'a', 'b', 'c' ]
   * pull(arr, 'a', 'c')
   * console.log(arr) // => [ 'b', 'b' ]
   */
  var values = slice.call(arguments, 1);
  for (var i = 0; i < values.length; i++) {
    var value = values[i];
    var fromIndex = -1;
    while ((fromIndex = indexOf.call(arr, value)) > -1) {
      splice.call(arr, fromIndex, 1);
    }
  }
  return arr;
}

function pullAt(arr, indexes) {
  if (!isArrayLike(arr)) {
    return [];
  }
  var length = arr ? indexes.length : 0;
  var last = length - 1;

  while (length--) {
    var previous = void 0;
    var index = indexes[length];
    if (length === last || index !== previous) {
      previous = index;
      splice.call(arr, index, 1);
    }
  }
  return arr;
}

function remove(arr, predicate) {
  /**
   * const arr = [1, 2, 3, 4]
   * const evens = remove(arr, n => n % 2 == 0)
   * console.log(arr) // => [1, 3]
   * console.log(evens) // => [2, 4]
   */
  var result = [];
  if (!isArrayLike(arr)) {
    return result;
  }
  var i = -1;
  var indexes = [];
  var length = arr.length;

  while (++i < length) {
    var value = arr[i];
    if (predicate(value, i, arr)) {
      result.push(value);
      indexes.push(i);
    }
  }
  pullAt(arr, indexes);
  return result;
}

var arrayUtil = {
  contains: contains,
  pull: pull,
  pullAll: pull,
  pullAt: pullAt,
  remove: remove,
  uniq: uniq
};

module.exports = arrayUtil;
