// umi-test/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _jestCli = _interopRequireDefault(require("jest-cli"));

var _path = require("path");

var _fs = require("fs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const debug = require('debug')('umi-test');

process.env.NODE_ENV = 'test';

function test(path) {
  return (0, _fs.existsSync)(path) && (0, _fs.statSync)(path).isDirectory();
}

function _default(opts = {}) {
  const watch = opts.watch,
        coverage = opts.coverage,
        _opts$libraryName = opts.libraryName,
        libraryName = _opts$libraryName === void 0 ? 'umi' : _opts$libraryName,
        _opts$cwd = opts.cwd,
        cwd = _opts$cwd === void 0 ? process.cwd() : _opts$cwd;
  const jestConfigFile = (0, _path.join)(cwd, 'jest.config.js');
  let userJestConfig = {};

  if ((0, _fs.existsSync)(jestConfigFile)) {
    userJestConfig = require(jestConfigFile); // eslint-disable-line
  }

  let pagesPath = 'pages';

  if (test((0, _path.join)(cwd, 'src/page'))) {
    pagesPath = 'src/page';
  }

  if (test((0, _path.join)(cwd, 'src/pages'))) {
    pagesPath = 'src/pages';
  }

  const config = _objectSpread({
    rootDir: process.cwd(),
    setupFiles: [require.resolve('./shim.js'), require.resolve('./setupTests.js')],
    transform: {
      '\\.jsx?$': require.resolve('./transformers/jsTransformer'),
      '\\.tsx?$': require.resolve('./transformers/tsTransformer')
    },
    testMatch: ['**/?(*.)(spec|test|e2e).(j|t)s?(x)'],
    moduleFileExtensions: ['js', 'jsx', 'ts', 'tsx', 'json'],
    setupTestFrameworkScriptFile: require.resolve('./jasmine'),
    moduleNameMapper: {
      '\\.(css|less|sass|scss)$': require.resolve('identity-obj-proxy')
    },
    globals: {
      'ts-jest': {
        useBabelrc: true
      }
    }
  }, coverage ? {
    collectCoverageFrom: ['pages/**/*.{ts,tsx,js,jsx}', 'src/**/*.{ts,tsx,js,jsx}', '!**/*.d.ts'],
    collectCoverage: true,
    coveragePathIgnorePatterns: [`/${pagesPath}/.${libraryName}/`, `/${pagesPath}/.${libraryName}-production/`]
  } : {}, userJestConfig || {});

  return new Promise((resolve, reject) => {
    _jestCli.default.runCLI({
      watch,
      testPathPattern: process.argv.slice(2).filter(arg => !arg.startsWith('-')),
      config: JSON.stringify(config)
    }, [cwd]).then(result => {
      debug(result);
      const results = result.results; // const success = results.every(result => result.success);

      if (results.success) {
        resolve();
      } else {
        reject(new Error('Jest failed'));
      }
    }).catch(e => {
      console.log(e);
    });
  });
}
